# Chapter 1: Business Understanding – Sentiment Analysis on Tech Products

## 1.1 Background Information

In today’s digital age, consumers constantly share their opinions online through reviews, tweets, and forums. These opinions hold valuable insights for technology companies that rely heavily on user experience and brand perception. Understanding public sentiment helps firms make data-driven product, marketing, and customer experience decisions.

This project is undertaken by Lingua Franca Data Elites, a data-driven consulting firm specializing in artificial intelligence solutions for business intelligence and brand analytics. Lingua Franca
 has been contracted to conduct a sentiment analysis on customer reviews and discussions surrounding Google and Apple products. With both brands dominating the global tech landscape, analyzing how customers perceive their products across online platforms can reveal comparative advantages, recurring complaints, and evolving brand loyalty trends.

The motivation for this analysis arises from the increasing competition in the consumer technology market. Both Google and Apple continuously release innovative hardware and software products, yet their customer satisfaction levels and public sentiment fluctuate based on product performance, pricing, design, and ecosystem integration. Understanding these sentiments provides an opportunity for firms and investors to forecast brand performance, guide marketing strategies, and identify product improvement areas.

--- 


## 1.2 Problem Statement

Despite the abundance of online reviews and discussions, companies often struggle to systematically analyze this data to extract actionable insights. Manual review analysis is time-consuming, subjective, and impractical given the volume of data generated daily.

This project aims to develop a Natural Language Processing (NLP) model that automatically classifies and interprets user sentiment (positive, negative, or neutral) toward Google and Apple products. By analyzing patterns and themes within the text data, the project seeks to answer key business questions such as:

* How do users perceive Google versus Apple products in terms of satisfaction and quality?
* What product features drive the strongest positive or negative emotions?
* How has sentiment evolved over time for each brand?
* Are there recurring issues or praise patterns that can inform future product development and customer engagement?

---

## 1.3 Project Objectives

### General Objective:

To build and evaluate an NLP model capable of accurately predicting and interpreting customer sentiment regarding Google and Apple products, providing actionable insights for strategic decision-making.

### Specific Objectives:

1. To perform data cleaning and exploratory data analysis to understand trends and key topics in user reviews.
2. To develop a baseline machine learning model (TF-IDF + Logistic Regression) for sentiment classification.
3. To fine-tune a transformer-based NLP model (e.g., BERT or DistilBERT) for improved sentiment prediction accuracy.
4. To conduct error and interpretability analysis to understand which words or phrases drive sentiment outcomes.
5. To compare sentiment distributions between Google and Apple products and highlight areas of strength or improvement for each brand.
6. To visualize sentiment trends and generate a dashboard/report for business stakeholders.

---

##  1.4 Metrics of Success

The success of this sentiment analysis project will be evaluated using both **quantitative model performance metrics** and **qualitative business outcomes**, each with clear numerical targets to guide model development and evaluation.

---

### a) Model Performance Metrics

| **Metric** | **Definition / Purpose** | **Baseline Target (TF-IDF + Logistic Regression)** | **Advanced Model Target (DistilBERT/BERT)** | **Business Interpretation** |
|-------------|---------------------------|----------------------------------------------------|---------------------------------------------|-----------------------------|
| **Accuracy** | Proportion of correctly classified reviews. | ≥ **75%** | ≥ **85–90%** | Reflects overall reliability in labeling sentiments correctly. |
| **Precision (Positive Class)** | Out of all predicted positive reviews, how many are actually positive. | ≥ **70%** | ≥ **85%** | High precision means fewer false positives — ideal for confident marketing insights. |
| **Recall (Negative Class)** | Out of all actual negative reviews, how many are detected by the model. | ≥ **70%** | ≥ **85%** | Ensures that major customer complaints are not missed. |
| **Macro F1-Score** *(Primary Metric)* | Harmonic mean of precision & recall across all classes. | ≥ **72%** | ≥ **87–90%** | Balanced measure ensuring fair performance across positive, neutral, and negative sentiments. |
| **Confusion Matrix** | Distribution of predictions per class. | Misclassification < **25%** overall | Misclassification < **10–15%** overall | Helps diagnose bias toward dominant sentiment classes. |
| **ROC-AUC (Binary View)** | Area under the ROC curve (if binary polarity used). | ≥ **0.80** | ≥ **0.92** | Measures the model’s ability to distinguish positive vs. negative classes. |

---

### b) Business-Level Success Indicators

| **Indicator** | **Target** | **Interpretation** |
|----------------|------------|--------------------|
| **Sentiment distribution clarity** | ≥ **70%** of reviews confidently classified (probability > 0.8) | Ensures reliable and actionable sentiment breakdowns. |
| **Insight generation** | Identify **top 5 drivers** of positive and negative sentiment. | Provides concrete recommendations for product and marketing teams. |
| **Visualization quality** | Dashboard with ≥ **3 key visual layers** (trend, brand comparison, key terms). | Demonstrates interpretability and usability of findings for business stakeholders. |

---

### c) Success Threshold Summary

- **Acceptable:** Macro F1 ≥ **0.72**, Accuracy ≥ **75%** (Baseline)  
- **Target:** Macro F1 ≥ **0.87**, Accuracy ≥ **85%** (DistilBERT)  
- **Excellent:** Macro F1 ≥ **0.90**, Accuracy ≥ **88–90%**, with actionable sentiment insights and clear interpretability visuals  

---

## 1.5 Expected Insights

By the end of this project, Lingua Franca Data Elites  expects to uncover meaningful insights that can influence both marketing and product strategies, including:

* Comparative sentiment analysis between Google and Apple, highlighting consumer preference patterns.

* Key drivers of satisfaction (e.g., design, usability, ecosystem) versus frustration (e.g., price, battery, compatibility).

* Temporal trends showing how major product launches or updates affect public opinion.

* Actionable recommendations for brand positioning and product improvement initiatives.


# Chapter 2: Data Understanding

In [4]:
# Importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from wordcloud import WordCloud
import re
import warnings
warnings.filterwarnings('ignore')

# Importing sklearn libraries
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler, RobustScaler, MinMaxScaler, LabelEncoder
from sklearn.model_selection import train_test_split, cross_val_score, cross_validate, GridSearchCV, RandomizedSearchCV
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, confusion_matrix, classification_report,\
      precision_score, roc_curve, ConfusionMatrixDisplay 
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC,LinearSVC
from sklearn.naive_bayes import MultinomialNB



# Importing nltk 
import nltk
from nltk.tokenize import word_tokenize
nltk.download('stopwords')
nltk.download('punkt')
from nltk.corpus import stopwords
from nltk.stem import SnowballStemmer



[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\user\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\user\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [ ]:
# Loading the dataset
data = pd.read_csv('judge-1377884607_tweet_product_company.csv', encoding = 'latin-1')

# Previewing the first five rows
data.head()

,tweet_text,emotion_in_tweet_is_directed_at,is_there_an_emotion_directed_at_a_brand_or_product
0,.@wesley83 I have a 3G iPhone. After 3 hrs twe...,iPhone,Negative emotion
1,@jessedee Know about @fludapp ? Awesome iPad/i...,iPad or iPhone App,Positive emotion
2,@swonderlin Can not wait for #iPad 2 also. The...,iPad,Positive emotion
3,@sxsw I hope this year's festival isn't as cra...,iPad or iPhone App,Negative emotion
4,@sxtxstate great stuff on Fri #SXSW: Marissa M...,Google,Positive emotion


In [ ]:
# Previewing the last five rows
data.tail()

,tweet_text,emotion_in_tweet_is_directed_at,is_there_an_emotion_directed_at_a_brand_or_product
9088,Ipad everywhere. #SXSW {link},iPad,Positive emotion
9089,"Wave, buzz... RT @mention We interrupt your re...",NaN,No emotion toward brand or product
9090,"Google's Zeiger, a physician never reported po...",NaN,No emotion toward brand or product
9091,Some Verizon iPhone customers complained their...,NaN,No emotion toward brand or product
9092,ŒÏ¡ŽÏàŠü_‹Ê‹Î‹Ò‹£‹Á‹â⋁_‹£‹‹â_‹ÛâRT @...,NaN,No emotion toward brand or product


Observation: The dataset is uniform from top to bottom.

In [11]:
# Checking the data types
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9093 entries, 0 to 9092
Data columns (total 3 columns):
 #   Column                                              Non-Null Count  Dtype 
---  ------                                              --------------  ----- 
 0   tweet_text                                          9092 non-null   object
 1   emotion_in_tweet_is_directed_at                     3291 non-null   object
 2   is_there_an_emotion_directed_at_a_brand_or_product  9093 non-null   object
dtypes: object(3)
memory usage: 213.2+ KB


Observations: All the columns in the dataset are categorical.

In [14]:
# Checking the shape
print(f" The dataset has {data.shape[0]} records and {data.shape[1]} features")

 The dataset has 9093 records and 3 features


In [16]:
# Checking statistical summary
data.describe().T

,count,unique,top,freq
tweet_text,9092,9065,RT @mention Marissa Mayer: Google Will Connect...,5
emotion_in_tweet_is_directed_at,3291,9,iPad,946
is_there_an_emotion_directed_at_a_brand_or_product,9093,4,No emotion toward brand or product,5389


Observations: 
* There are 9065 unique tweets. There may be need to check for duplicates.
* There are 9 unique products that the emotions are directed at. There may be need to feature engineer tojust have 2 categories either Apple or Google.
* There are 4 unique sentiments. However, they are meant to be 3 therefore there may be need to clean the column and standardize the text.

# Chapter 3: Data Preparation

In [17]:
# Checking for null values
data.isna().sum()

tweet_text                                               1
emotion_in_tweet_is_directed_at                       5802
is_there_an_emotion_directed_at_a_brand_or_product       0
dtype: int64

Observations: There is one null value in tweet_text and 5802 null values in emotion_in_tweet_is_directed_at columns which will be handled in data cleaning.

In [19]:
#Checking for duplicates
data.duplicated().sum()

np.int64(22)

Observations: There are 22 duplicated rows in the dataset.